src/Clapi/TextSerialisation.hs

```diff
@@ -24,10 +30,49 @@ tpBuilder ((Time s f), vs) = tb <> mconcat (map (\v -> (fromChar ' ') <> (cvBuil
   where
     tb = (fss s) <> (fromChar ':') <> (fss f)
 
-encode :: Map.Map Time [ClapiValue] -> Builder
+type TimeSeries = Map.Map Time [ClapiValue]
+
+encode :: TimeSeries -> Builder
 encode ts = header <> bodyBuilder
   where
     -- It is invalid for a time series to be empty, so this use of head is
     -- kinda fine, but the errors will suck:
     header = fromString $ map typeTag $ head $ Map.elems ts
     bodyBuilder = mconcat $ map (\tp -> fromChar '\n' <> tpBuilder tp) (Map.toList ts)
+
+quotedString :: Parser Text
+-- FIXME: escaped quotes unsupported
+quotedString = char '"' >> takeTill (== '"')
+
+cvParser :: Char -> Parser ClapiValue
+cvParser 'B' = (wordMatch 'T' True) <|> (wordMatch 'F' False)
+  where
+    wordMatch c v = char c >> return (ClBool v)
+cvParser 'i' = decimal >>= return . ClInt32
+cvParser 's' = ClString <$> quotedString
+
+getTupleParser :: Parser (Parser [ClapiValue])
+getTupleParser = sequence <$> many1 (satisfy (inClass "Bis") >>= \c -> return (char ' ' >> cvParser c))
+
+timeParser :: Parser Time
+timeParser = do
+    s <- decimal
+    char ':'
+    f <- decimal
+    return $ Time s f
+
+decode :: Text -> Either String TimeSeries
+decode txt = case parse getTupleParser txt of
+    Fail _ ctxs msg -> Left msg
```

Comment:
Did you break apart the parser state here so that you could writer better error messages, or because you're heading towards partial parsing, or something else? I don't recall needing to do it ever, but I think I just used `parseOnly` in a lot of places...

Reply:
I am wondering about effectively being able to stream in the file, but I haven't gone there yet. Mainly it's because parseOnly screwed me over before by not giving me enough to go on.

Reply:
Because this is a 2 stage thing at the moment I think it might also help with the errors (but I think it could be reworked to look 1 stagey).